### PR TITLE
[FIX] base: Validate URLs on attachments

### PR DESCRIPTION
**Pull request by ryce-odoo**
Added a validation for URLs on ir.attachments to resolve an issue where incomplete links on attachments will redirect to a local URL instead of an external URL. Thus leading to a page that does not exist.

e.g.: youtube.com instead of https://www.youtube.com

This seemed like the proper course of action as there would be no reason to not validate as any URL within the database can still be formatted properly and work the same.

opw-3698591
---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/addons/base/models/ir_attachment.py [hunk 1]
```diff
@@ -435,6 +435,14 @@ def _check_serving_attachments(self):
                 if not any(has_group(g) for g in attachment.get_serving_groups()):
                     raise ValidationError("Sorry, you are not allowed to write on this document")
 
+    @api.constrains('url')
+    def _check_url_validity(self):
+        for attachment in self:
+            if attachment.type == 'url' and attachment.url:
+                # Make sure that any URL that is inputted is a valid URL if the type is url
+                if not re.match(r"^(ht|f)tp(s?)\:\/\/[0-9a-zA-Z]([-.\w]*[0-9a-zA-Z])*(:(0-9)*)*(\/?)([a-zA-Z0-9\-\.\?\,\'\/\\\+&amp;%\$#_]*)?$", attachment.url):
+                    raise ValidationError(_("Please enter a valid URL \nExample: https://www.odoo.com"))
```
Comment (ryv-odoo) [inline odoo/addons/base/models/ir_attachment.py:444]: Please add the problematic url in the message ? Some form views can maybe contains several url linked to a ir.attachment

**Review by ryv-odoo — COMMENTED** [review]
Hi @ryce-odoo , it looks like your runbot is red (for at least one related test).
The idea seems relevant, but I am not sure if we should make this change in 15.0 though (some external module might be broken with it).
Also, should you add a test to cover the expected behavior?

File: odoo/addons/base/models/ir_attachment.py [hunk 2]
```diff
@@ -435,6 +435,14 @@ def _check_serving_attachments(self):
                 if not any(has_group(g) for g in attachment.get_serving_groups()):
                     raise ValidationError("Sorry, you are not allowed to write on this document")
 
+    @api.constrains('url')
+    def _check_url_validity(self):
+        for attachment in self:
+            if attachment.type == 'url' and attachment.url:
+                # Make sure that any URL that is inputted is a valid URL if the type is url
+                if not re.match(r"^(ht|f)tp(s?)\:\/\/[0-9a-zA-Z]([-.\w]*[0-9a-zA-Z])*(:(0-9)*)*(\/?)([a-zA-Z0-9\-\.\?\,\'\/\\\+&amp;%\$#_]*)?$", attachment.url):
```
[the remark this answers is not in the document]
Comment (ryce-odoo) [reply odoo/addons/base/models/ir_attachment.py:443]: Hey @ryv-odoo @xmo-odoo 
Thanks for the suggestions. I just pulled that regex from here: https://learn.microsoft.com/en-us/previous-versions/msp-n-p/ff650303(v=pandp.10)?redirectedfrom=MSDN#common-regular-expressions

I agree that it is quite overkill. I will look into changing that likely an http(s):// check should work fine in most cases.

File: odoo/addons/base/models/ir_attachment.py (hunk 1 above)
Comment (ryce-odoo) [reply odoo/addons/base/models/ir_attachment.py:444]: Noted

**Comment by ryce-odoo** [comment]
> Hi @ryce-odoo , it looks like your runbot is red (for at least one related test). The idea seems relevant, but I am not sure if we should make this change in 15.0 though (some external module might be broken with it). Also, should you add a test to cover the expected behavior?

Hey @ryv-odoo,
It seems like the tests are failing because there are some data files that use the relative URL functionality of ir.attachments. My proposed solution for this is to add a boolean field to the ir.attachments for external URLs and only trigger this constraint when that boolean is marked true. What are your thoughts on this?

**Comment by ryv-odoo** [comment]
> I see I would agree that it is overkill, but it seems that there are not any options in this case then. The issue with adding it to product.document is that this model only exists in 17.0 and the pertaining bugfix ticket has a customer on 16.0.

My bad, in 16.0 we modify the ir.attachment record. You can add a warning in the onchange (api.onchange) for 15.0/16.0 version, but it won't constraint the user to save it (also I may break some other untested flows where bad url are legit).

> Would it be sufficient to let them know that this is a limitation that we have solved in 17?

I think it is the best option, it is not a blocking issue anyway, just a nice to have.

**Comment by ryce-odoo** [comment]
Re-targeting to 17.0 
new pr: https://github.com/odoo/odoo/pull/155806

**Pull request closed without merging**